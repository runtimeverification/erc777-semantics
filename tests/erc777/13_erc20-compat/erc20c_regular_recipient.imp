test.preamble();

/*
Operations in each test:
    - transfer
    - transfer, more than balance
    - transfer, 0 amount
    - transferFrom from same address, enough allowance
    - transferFrom from different address, enough allowance
    - transferFrom from different address, more than allowance.

regular recipient, no hooks.*/

test.makeCaller(10);
test.transfer(1, 1);            // regular transfer, 1 is regular recipient
test.transfer(1, 0);            // 0 amount

test.approve(10, 20);
test.transferFrom(10, 1, 10);   // transferFrom from same address, enough allowance

test.approve(8, 20);
test.makeCaller(8);
test.transferFrom(10, 1, 10);   // from different address, enough allowance
test.transferFrom(10, 1, 11);   // from different address, more than allowance

test.makeCaller(10);
test.approve(8, 200);
test.makeCaller(8);
test.transferFrom(10, 1, 200);  // more than balance, enough allowance

printLogs;
