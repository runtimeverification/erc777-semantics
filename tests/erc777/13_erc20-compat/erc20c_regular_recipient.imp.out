Changing the caller to 10
10 is transferring 1 to 1
Balance of 10 is 99
Balance of 1 is 11
Allowance of 10 for 1 is 0
10 is transferring 0 to 1
Balance of 10 is 99
Balance of 1 is 11
Allowance of 10 for 1 is 0
10 is approving allowance 20 for 10
10 is transferring 10 from 10 to 1
Balance of 10 is 89
Balance of 1 is 21
Allowance of 10 for 10 is 10
10 is approving allowance 20 for 8
Changing the caller to 8
8 is transferring 10 from 10 to 1
Balance of 10 is 79
Balance of 1 is 31
Allowance of 10 for 8 is 10
8 is transferring 11 from 10 to 1
REVERT
Balance of 10 is 79
Balance of 1 is 31
Allowance of 10 for 8 is 10
Changing the caller to 10
10 is approving allowance 200 for 8
Changing the caller to 8
8 is transferring 200 from 10 to 1
REVERT
Balance of 10 is 79
Balance of 1 is 31
Allowance of 10 for 8 is 200

Events:
Minted(8, 10, 100, , init)
Transfer(0, 10, 100)
Minted(8, 9, 90, , init)
Transfer(0, 9, 90)
Minted(8, 8, 80, , init)
Transfer(0, 8, 80)
Minted(8, 7, 70, , init)
Transfer(0, 7, 70)
Minted(8, 6, 60, , init)
Transfer(0, 6, 60)
Minted(8, 5, 50, , init)
Transfer(0, 5, 50)
Minted(8, 4, 40, , init)
Transfer(0, 4, 40)
Minted(8, 3, 30, , init)
Transfer(0, 3, 30)
Minted(8, 2, 20, , init)
Transfer(0, 2, 20)
Minted(8, 1, 10, , init)
Transfer(0, 1, 10)
Transfer(10, 1, 1)
Sent(10, 10, 1, 1, , )
Transfer(10, 1, 0)
Sent(10, 10, 1, 0, , )
Approval(10, 10, 20)
Transfer(10, 1, 10)
Sent(10, 10, 1, 10, , )
Approval(10, 8, 20)
Transfer(10, 1, 10)
Sent(10, 10, 1, 10, , )
Approval(10, 8, 200)
